cylc/flow/task_pool.py

```diff
@@ -1378,32 +1347,51 @@ def remove_if_complete(self, itask):
             else (failed):
                 - retain and recompute runahead
                   (C7 failed tasks don't count toward runahead limit)
+
         """
-        ret = False
+        if not itask.state(*TASK_STATUSES_FINAL):
+            # can't be complete
+            return False
+
+        if itask.identity == self.stop_task_id:
+            self.stop_task_finished = True
+
         if cylc.flow.flags.cylc7_back_compat:
+            ret = False
             if not itask.state(TASK_STATUS_FAILED, TASK_OUTPUT_SUBMIT_FAILED):
-                self.remove(itask, 'finished')
+                self.remove(itask)
                 ret = True
+            # Recompute runahead either way; failed tasks don't count in C7.
             if self.compute_runahead():
                 self.release_runahead_tasks()
-        else:
-            incomplete = itask.state.outputs.get_incomplete()
-            if incomplete:
-                # Retain as incomplete.
+            return ret
+
+        if itask.state(TASK_STATUS_EXPIRED):
+            self.remove(itask, "expired")
+            if self.compute_runahead():
+                self.release_runahead_tasks()
+            return True
+
+        incomplete = itask.state.outputs.get_incomplete()
+        if incomplete:
+            # Keep incomplete tasks in the pool.
+            if output in (
+                TASK_STATUS_SUCCEEDED, TASK_STATUS_FAILED, TASK_STATUS_EXPIRED,
+                TASK_STATUS_SUBMIT_FAILED
+            ):
```

Comment:
We've got a constant for this which is already imported:

```suggestion
            if output in TASK_STATUSES_FINAL:
```